### PR TITLE
Add ability to specify storage location

### LINKED ISSUE #43: Option to specify alternate location for local zstash files
Currently, zstash will store local copies of database and tar files under the sub-directory `zstash/`.

In certain cases (see for example https://github.com/E3SM-Project/zstash/issues/41), it might be useful to have the flexibility to specify and alternate location, maybe with a new command line argument:

`--cache=...`

Is this worth considering implementing? Or would it just add unnecessary confusion?

### PR DESCRIPTION
**Pull request by forsyth2**
This resolves #43.

- [x] Switch `CACHE` to `config.cache` to make storage location easy to change.
- [x] Add `--cache` option.
- [x] Add tests of new option.

### CODE REVIEW

**Comment by forsyth2** [comment]
@golaz I added the `--cache` option to `chgrp`, `create`, `extract`, `ls`, and `update`. Or should it only be an option to `create`? Testing will certainly get more complicated if `config.cache` can be changed as part of any operation.

**Comment by golaz** [comment]
Yes, we will need it as well for create. Is there a way we can restructure the tests so we can loop over all combination of options rather than repeating tests?

File: zstash/chgrp.py [hunk 1]
```diff
@@ -15,12 +15,18 @@ def chgrp():
...
     parser.add_argument('-v', '--verbose', action="store_true", 
                         help="increase output verbosity")
 
     args = parser.parse_args(sys.argv[2:])
     if args.hpss and args.hpss.lower() == 'none':
         args.hpss = 'none'
+    if args.cache:
+        config.cache = args.cache
+    else:
+        config.cache = 'zstash'
```
Comment (forsyth2) [inline zstash/chgrp.py]: `zstash_cache`

File: zstash/create.py [hunk 2]
```diff
@@ -51,6 +53,10 @@ def create():
     config.hpss = args.hpss
     config.maxsize = int(1024*1024*1024 * args.maxsize)
     config.keep = args.keep
+    if args.cache:
+        config.cache = args.cache
+    else:
+        config.cache = 'zstash'
```
Comment (forsyth2) [inline zstash/create.py]: `zstash_cache`

File: zstash/extract.py [hunk 3]
```diff
@@ -103,31 +103,37 @@ def extract(keep_files=True):
...
     optional.add_argument('-v', '--verbose', action="store_true", 
                           help="increase output verbosity")
     parser.add_argument('files', nargs='*', default=['*'])
     args = parser.parse_args(sys.argv[2:])
     if args.hpss and args.hpss.lower() == 'none':
         args.hpss = 'none'
+    if args.cache:
+        config.cache = args.cache
+    else:
+        config.cache = 'zstash'
```
Comment (forsyth2) [inline zstash/extract.py]: `zstash_cache`

File: zstash/ls.py [hunk 4]
```diff
@@ -21,29 +21,35 @@ def ls():
...
                           help="increase output verbosity")
     
     parser.add_argument('files', nargs='*', default=['*'])
     args = parser.parse_args(sys.argv[2:])
     if args.hpss and args.hpss.lower() == 'none':
         args.hpss = 'none'
+    if args.cache:
+        config.cache = args.cache
+    else:
+        config.cache = 'zstash'
```
Comment (forsyth2) [inline zstash/ls.py]: `zstash_cache`

File: zstash/update.py [hunk 5]
```diff
@@ -32,26 +32,32 @@ def update():
...
+        '--cache', type=str, help='path to store files')
     optional.add_argument('-v', '--verbose', action="store_true", 
                           help="increase output verbosity")
     args = parser.parse_args(sys.argv[2:])
     if args.hpss and args.hpss.lower() == 'none':
         args.hpss = 'none'
+    if args.cache:
+        config.cache = args.cache
+    else:
+        config.cache = 'zstash'
```
Comment (forsyth2) [inline zstash/update.py]: `zstash_cache`

**Comment by forsyth2** [comment]
@golaz ready for code review

File: tests/test_check.py [hunk 6]
```diff
@@ -13,33 +13,38 @@ class TestCheck(TestZstash):
     # option | Check | CheckMismatch | CheckKeepTars | CheckParallel | CheckParallelVerboseMismatch | CheckParallelKeepTars |
     # --hpss    |x|x|x|x|x| |
     # --workers | | | |x|x|x|
+    # --cache   |x| | | | | |
```
Comment (forsyth2) [inline tests/test_check.py:16]: Also update for `test_check_parallel.py`

File: tests/test_extract.py [hunk 7]
```diff
@@ -10,11 +10,12 @@ class TestExtract(TestZstash):
...
     # x = on, no mark = off, b = both on and off tested
-    # option | ExtractVerbose | ExtractKeep | ExtractParallel |
-    # --hpss    |x|x|x|
-    # --workers | | |x|
-    # --keep    | |x| |
-    # -v        |x| |b|
+    # option | ExtractVerbose | ExtractKeep | ExtractCache | ExtractParallel |
+    # --hpss    |x|x|x|x|
+    # --workers | | | |x|
+    # --cache   | | |x| |
```
Comment (forsyth2) [inline tests/test_extract.py:16]: Also update for `test_extract_parallel.py`

**Comment by golaz** [comment]
@forsyth2 : I checked out the code and tried running the tests. But I'm getting a lot of test failures:

```
Ran 46 tests in 180.111s

FAILED (failures=22, errors=14)
```

Not sure if I did something wrong. Here is what I ran on cori:

```
module load python/3.7-anaconda-2019.10
source /global/common/cori_cle7/software/python/3.7-anaconda-2019.10/etc/profile.d/conda.sh
conda activate zstash_env_dev
cd E3SM/code
rm -rf zstash
git clone https://github.com/forsyth2/zstash.git
cd zstash
git checkout alt-location
python setup.py install
cd tests
python -m unittest
```

**Comment by forsyth2** [comment]
@golaz All 46 tests pass for me on Cori. I run with:
```
cd /global/cscratch1/sd/forsyth/zstash/tests
conda activate zstash_env41
pip install .. && python -m unittest
```
Maybe our conda environments are different? Or maybe the `setup.py install` does something different?

Also, for future reference, you don't need to delete your `zstash` repo to use my branch:
```
git remote -v
# Add my repo as a remote if you don't see it listed.
# The copy-link is obtained from the "Clone or Download" button on https://github.com/forsyth2/zstash
git remote add forsyth2 git@github.com:forsyth2/zstash.git
git remote -v
# You should now see my repo as a remote.
git fetch forsyth2 alt-location
git checkout -b test-forsyth2-alt-location forsyth2/alt-location
# You should now be on a branch identical to mine.
```

**Comment by golaz** [comment]
@forsyth2 : I re-installed with pip. Unfortunately, I'm still getting test failures, but fewer. The failuers appear to be all similar:

```
======================================================================
FAIL: testCheckVerboseMismatchHPSS (test_check.TestCheck)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check.py", line 97, in testCheckVerboseMismatchHPSS
    self.helperCheckVerboseMismatch('testCheckVerboseMismatchHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check.py", line 54, in helperCheckVerboseMismatch
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testCheckParallelHPSS (test_check_parallel.TestCheckParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 139, in testCheckParallelHPSS
    self.helperCheckParallel('testCheckParallelHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 25, in helperCheckParallel
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testCheckParallelVerboseHPSS (test_check_parallel.TestCheckParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 132, in testCheckParallelVerboseHPSS
    self.helperCheckParallel('testCheckParallelVerboseHPSS', HPSS_ARCHIVE, verbose=True)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 25, in helperCheckParallel
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testCheckParallelVerboseMismatchHPSS (test_check_parallel.TestCheckParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 146, in testCheckParallelVerboseMismatchHPSS
    self.helperCheckParallelVerboseMismatch('testCheckParallelVerboseMismatchHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_check_parallel.py", line 47, in helperCheckParallelVerboseMismatch
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testExtractCacheHPSS (test_extract.TestExtract)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 153, in testExtractCacheHPSS
    self.helperExtractCache('testExtractCacheHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 119, in helperExtractCache
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testExtractKeepHPSS (test_extract.TestExtract)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 146, in testExtractKeepHPSS
    self.helperExtractKeep('testExtractKeepHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 83, in helperExtractKeep
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testExtractVerboseHPSS (test_extract.TestExtract)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 139, in testExtractVerboseHPSS
    self.helperExtractVerbose('testExtractVerboseHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract.py", line 25, in helperExtractVerbose
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

======================================================================
FAIL: testExtractParallelHPSS (test_extract_parallel.TestExtractParallel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract_parallel.py", line 86, in testExtractParallelHPSS
    self.helperExtractParallel('testExtractParallelHPSS', HPSS_ARCHIVE)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/test_extract_parallel.py", line 25, in helperExtractParallel
    use_hpss = self.setupDirs(test_name)
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 172, in setupDirs
    self.assertWorkspace()
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 134, in assertWorkspace
    self.stop('Current_directory={} does not end in {}'.format(actual, expected))
  File "/global/u2/g/golaz/E3SM/code/zstash/tests/base.py", line 124, in stop
    self.fail(error_message)
AssertionError: Current_directory=/global/u2/g/golaz/E3SM/code/zstash/tests/zstash_test does not end in zstash/tests

----------------------------------------------------------------------
Ran 46 tests in 391.705s

FAILED (failures=8, errors=8)

```

**Comment by forsyth2** [comment]
@golaz Ah, I added the directory check to make sure the tests were running from the correct directory (and thus, wouldn't delete the repo `zstash` instead of the cache `zstash`, after that happened once).

Are you running `python -m unittests` from a directory `zstash/tests/zstash_test`? In that case, try running from `zstash/tests`.

**Comment by golaz** [comment]
No, I was running from within

`zstash/tests`

**Comment by golaz** [comment]
@forsyth2 : I can confirm that all the tests pass if I run them from $CSCRATCH

```
Ran 46 tests in 976.638s

OK
```

I wonder whether we should somehow enforce that, or warn users against running from /global/homes.

**Comment by golaz** [comment]
@forsyth2 : I have now run a few more tests to familiarize myself with the new option. It works well and feels very intuitive to me. I could see this option as being very useful, for example if we needed to maintain a central read-only zstash archive that multiple users could share.

I have only one minor suggestion for consideration: I would have preferred for the default cache name to remain 'zstash' instead of 'zstash_cache'. I know it's confusing to have so many things called 'zstash', but I kind of liked the simplicity of the name. Plus, the default name change could also confuse users and might give the impression of "breaking" things if one user upgrades to the latest version of zstash and tries to access some older archive where there is already a zstash sub-directory present.

**Comment by forsyth2** [comment]
@golaz Thanks for the review!

I'll add a comment in the test to warn users against running from `/global/homes`. Maybe I should also update the docs to mention that.

I'll change the default cache back to `zstash`. We (developers and users) just need to be extra careful we know whether we're working with the repo `zstash`, the cache `zstash`, or even the command `zstash`.

**Comment by golaz** [comment]
@forsyth2 : great work! Please go ahead and merge it to master.

**Pull request merged**